chalk-ir/src/lib.rs

```diff
@@ -1227,17 +1227,42 @@ impl<I: Interner> ProgramClauseImplication<I> {
     }
 }
 
-impl<I: Interner> ProgramClause<I> {
-    pub fn into_from_env_clause(self, interner: &I) -> ProgramClause<I> {
+impl<I: Interner> ProgramClauseData<I> {
+    pub fn into_from_env_clause(self, interner: &I) -> ProgramClauseData<I> {
         match self {
-            ProgramClause::Implies(implication) => {
-                ProgramClause::Implies(implication.into_from_env_clause(interner))
+            ProgramClauseData::Implies(implication) => {
+                ProgramClauseData::Implies(implication.into_from_env_clause(interner))
             }
-            ProgramClause::ForAll(binders_implication) => {
-                ProgramClause::ForAll(binders_implication.map(|i| i.into_from_env_clause(interner)))
+            ProgramClauseData::ForAll(binders_implication) => {
+                ProgramClauseData::ForAll(binders_implication.map(|i| i.into_from_env_clause(interner)))
             }
         }
     }
+
+    pub fn intern(self, interner: &I) -> ProgramClause<I> {
+        ProgramClause {
+            clause: interner.intern_program_clause(self),
+        }
+    }
+}
+
+#[derive(Clone, PartialEq, Eq, Hash, PartialOrd, Ord, HasInterner)]
+pub struct ProgramClause<I: Interner> {
+    clause: I::InternedProgramClause,
```

Comment:
Nit: I think we we call this `interned` in all the other structs

Reply:
We call it `interned` except in `Goal` and `Goals` IIRC. I think those should be renamed and everything should be `interned`.

Reply:
It looks like @tirr-c fixed the new cases in this PR, we can create an issue for the others